src/dsl/fusion/GroupFusion.hpp

```diff
@@ -0,0 +1,98 @@
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef NUCLEAR_DSL_FUSION_GROUPFUSION_HPP
+#define NUCLEAR_DSL_FUSION_GROUPFUSION_HPP
+
+#include <algorithm>
+
+#include "../../threading/Reaction.hpp"
+#include "../operation/DSLProxy.hpp"
+#include "has_group.hpp"
+
+namespace NUClear {
+namespace dsl {
+    namespace fusion {
+
+        /// Type that redirects types without a group function to their proxy type
+        template <typename Word>
+        struct Group {
+            using type = std::conditional_t<has_group<Word>::value, Word, operation::DSLProxy<Word>>;
+        };
+
+        template <typename, typename = std::tuple<>>
+        struct GroupWords;
+
+        /**
+         * @brief Metafunction that extracts all of the Words with a group function
+         *
+         * @tparam Word1        The word we are looking at
+         * @tparam WordN        The words we have yet to look at
+         * @tparam FoundWords   The words we have found with group functions
+         */
+        template <typename Word1, typename... WordN, typename... FoundWords>
+        struct GroupWords<std::tuple<Word1, WordN...>, std::tuple<FoundWords...>>
+            : public std::conditional_t<
+                  has_group<typename Group<Word1>::type>::value,
+                  /*T*/ GroupWords<std::tuple<WordN...>, std::tuple<FoundWords..., typename Group<Word1>::type>>,
+                  /*F*/ GroupWords<std::tuple<WordN...>, std::tuple<FoundWords...>>> {};
+
+        /**
+         * @brief Termination case for the GroupWords metafunction
+         *
+         * @tparam FoundWords The words we have found with group functions
+         */
+        template <typename... FoundWords>
+        struct GroupWords<std::tuple<>, std::tuple<FoundWords...>> {
+            using type = std::tuple<FoundWords...>;
+        };
+
+
+        // Default case where there are no group words
+        template <typename Words>
+        struct GroupFuser {};
+
+        // Case where there is only a single word remaining
+        template <typename Word>
+        struct GroupFuser<std::tuple<Word>> {
+
+            template <typename DSL>
+            static inline util::GroupDescriptor group(threading::Reaction& reaction) {
+
+                // Return our group
+                return Word::template group<DSL>(reaction);
+            }
+        };
+
+        // Case where there are 2 or more words remaining
+        template <typename Word1, typename Word2, typename... WordN>
+        struct GroupFuser<std::tuple<Word1, Word2, WordN...>> {
+
+            template <typename DSL>
+            static inline void group(threading::Reaction& /*reaction*/) {
+                throw std::runtime_error("Can not be a member of more than one group");
```

Comment:
Did we want to be able to be a part of multiple groups?

Reply:
Im not a fan of the idea, I can't visualise a viable use case for it, but if you are insistent and there are no other outstanding issues we can add that in a new PR